File: lessons/ads/queue.py
from typing import Any


class NodeD:
    def __init__(self, v: Any) -> None:
        self.value = v
        self._next = None
        self._prev = None

    def __setattr__(self, __name: str, __value: "NodeD") -> None:
        if __name == "next":
            self._next = __value
        if __name == "prev":
            self._prev = __value
        else:
            return super().__setattr__(__name, __value)

    def __getattribute__(self, __name: str) -> Any:
        if __name == "next":
            return self._next if not isinstance(self._next, DummyNode) else None
        if __name == "prev":
            return self._prev if not isinstance(self._prev, DummyNode) else None
        return super().__getattribute__(__name)


class DummyNode(NodeD):
    def __init__(self) -> None:
        super().__init__("DUMMY")


class DummyLinkedList:
    def __init__(self) -> None:
        self._head = DummyNode()
        self._tail = DummyNode()
        self._head._next = self._tail
        self._tail._prev = self._head
        self._head._prev = self._tail._next = None

    def __setattr__(self, __name: str, __value: NodeD) -> None:
        if __name == "head":
            __value._prev = self._head
            __value._prev._next = __value
        elif __name == "tail":
            __value._next = self._tail
            __value._next._prev = __value
        else:
            return super().__setattr__(__name, __value)

    def __getattribute__(self, __name: str) -> Any:
        if __name == "head":
            return self._head._next if self._head._next is not self._tail else None
        if __name == "tail":
            return self._tail._prev if self._tail._prev is not self._head else None
        return super().__getattribute__(__name)

    def add_in_tail(self, value: Any) -> None:
        item = NodeD(value)
        item._next = self._tail
        item._prev = self._tail._prev
        item._prev._next = item
        item._next._prev = item

    def _delete(self, node: NodeD) -> None:
        node._next._prev = node._prev
        node._prev._next = node._next

    def pop_from_head(self) -> Any:
        node = self.head
        if node:
            self._delete(self.head)
            return node.value
        return None

    def len(self):
        counter = 0
        cur_node = self.head
        while cur_node:
            counter += 1
            cur_node = cur_node.next
        return counter


class Queue:
    def __init__(self) -> None:
        # инициализация хранилища данных
        self.queue = DummyLinkedList()
        self.counter = self.queue.len()

    def enqueue(self, item: Any) -> None:
        # вставка в хвост - O(1)
        self.queue.add_in_tail(item)
        self.counter += 1

    def dequeue(self) -> Any:
        # выдача из головы - O(1)
        self.counter -= 1
        return self.queue.pop_from_head()

    def size(self) -> int:
        return self.counter

File: lessons/ads3_go/Queue/Queue.go
// 5

package main

import (
	"errors"
	"os"
)

// Поскольку в слайсах операция добавления по индексу будет нам стоить O(n) на копирование слайса, то
// эффективнее воспользоваться связным списком для неограниченной очереди - тогда все операции 
// будут стоить O(1)
type Node[T any] struct {
	value T
	next  *Node[T]
	prev  *Node[T]
}

// 5.1.
type Queue[T any] struct {
	head *Node[T]
	tail *Node[T]
	size int
}

// 5.1.
func (q *Queue[T]) Size() int {
	return q.size
}

func (q *Queue[T]) getHead() T {
	return q.head.value
}

func (q *Queue[T]) removeHead() {
	newHead := q.head.next

	q.head.prev = nil
	q.head.next = nil
	q.head = newHead
	if q.head != nil {
		q.head.prev = nil
	} else {
		q.tail = nil
	}
	q.size--
}

// 5.1.
func (q *Queue[T]) Dequeue() (T, error) {
	var result T
	var zero T

	if q.size == 0 {
		return zero, errors.New("Empty queue")
	}

	result = q.getHead()
	q.removeHead()
	return result, nil
}

// 5.1.
func (q *Queue[T]) Enqueue(itm T) {
	n := &Node[T]{value: itm, prev: q.tail}

	if q.size == 0 {
		q.head = n
	} else {
		q.tail.next = n
	}
	q.tail = n
	q.size++
}

// 5.2. В этой реализации добавление и получение значения будет стоить O(1) для временной и пространственнй сложности
